In [1]:
import pandas as pd
import numpy as np

# **Урок 2. Анализ датасета с помощью Pandas**

## ***Задание №1***

### ***Скачать файл в уроке***
### ***Считать данные с помощью pandas***

In [2]:
df=pd.read_csv('kc_house_data.csv')

### ***Вывести на экран первые 5 строк***

In [3]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

### ***Посмотреть на описание признаков и на их содержание***

### Наименование разделов таблицы:
- id - Идентифкотор дома
- date - Дата сделки по продаже дома
- price - Стоимость продажи дома
- bedrooms - Количество спален
- bathrooms - Количество ванных комнат (0.5 - туалет без душа)
- sqft_living - Жилая пдлощадь (кв.футов)
- sqft_lot - Общая площадь (кв.футов)
- floors - Количество этажей
- waterfront - Вид на набережную
- view - Вид  - значение от 0 до 4
- condition - Состояние дома - значение от 1 до 5
- grade - Планировка и дизайн - значение от 1 до 13, где 1-3 плохая планировка здания и дизайн, 7 - средний уровень планировкии и дизайна, 11-13 - высокое качество планировки и дизайна
- sqft_above - Площадь здания выше уровня земли(без подвала)(кв. футов)
- sqft_basement - Площадь подвала (кв. футов)
- yr_built - Год постройки дома
- yr_renovated - Год проведения ремонта дома
- zipcode - Почтовый индекс
- lat - Широта
- long - Долгота 
- sqft_living15 - Жилая площадь  у 15 соседей (кв.футов)
- sqft_lot15 - Общая площадь у 15 соседей (кв.футов)

## ***Задание №2***

### ***Проведите первичный анализ данных***

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

### ***Изучите типы данных***

In [5]:
df.dtypes

id                 int64
date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

### ***Найдите количество пропущенных ячеек в данных***

In [6]:
df.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

### ***Посчитайте основные статистики по всем признакам и поизучайте их***

In [7]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [8]:
df.astype('object').describe()

id             date     price  bedrooms  bathrooms  \
count       21613            21613   21613.0     21613    21613.0   
unique      21436              372    4028.0        13       30.0   
top     795000620  20140623T000000  350000.0         3        2.5   
freq            3              142     172.0      9824     5380.0   

        sqft_living  sqft_lot   floors  waterfront   view  ...  grade  \
count         21613     21613  21613.0       21613  21613  ...  21613   
unique         1038      9782      6.0           2      5  ...     12   
top            1300      5000      1.0           0      0  ...      7   
freq            138       358  10680.0       21450  19489  ...   8981   

        sqft_above  sqft_basement  yr_built  yr_renovated  zipcode  \
count        21613          21613     21613         21613    21613   
unique         946            306       116            70       70   
top           1300              0      2014             0    98103   
freq           212          13126       559         20699      602   

               lat      long  sqft_living15  sqft_lot15  
count   21613.0000  21613.00          21613       21613  
unique   5034.0000    752.00            777        8689  
top        47.6624   -122.29           1540        5000  
freq       17.0000    116.00            197         427  

[4 rows x 21 columns]

In [9]:
df.describe(include=['object'])

date
count             21613
unique              372
top     20140623T000000
freq                142

## ***Задание №3***
### ***Ответьте на несколько вопросов***

### ***3.1 В каком диапазоне изменяются стоимости недвижимости?***

In [10]:
df['price'].min(), df['price'].max()

(75000.0, 7700000.0)

### ***3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?***

In [11]:
f"{round((df['sqft_living'].mean() * 100 / df['sqft_lot'].mean()), 2)} %"

'13.77 %'

### ***3.3 Как много домов с разными этажами в данных?***

In [12]:
df['floors'].value_counts()

1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: floors, dtype: int64

### ***3.4 Насколько хорошие состояния у домов в данных?***

In [13]:
df['condition'].value_counts()

3    14031
4     5679
5     1701
2      172
1       30
Name: condition, dtype: int64

### ***3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?***

In [14]:
f"Год строительтва первого дома  - {df['yr_built'].min()}"


'Год строительтва первого дома  - 1900'

In [15]:
f"Год строительства последнего дома - {df['yr_built'].max()}"

'Год строительства последнего дома - 2015'

## ***Выводы***

***База данных содержить 21613  записей с характеристиками объектов недвижимости, выставленных на продажу. Все записи содержать полную информацию.  Анализ проводится по 20 признакам.***



***Большую часть (10680 объектов) составляют одноэтажные дома, пользующиеся повышенным спросом. На втором месте по количеству этажей 2-х этажные  - 8241 объектов. Спрос на данный вид объектов стоит на втором месте. Имеются объекты с большей этажностью, а также с наличием подвальных помещений.***  

***Стоимотсь объектов представлена в диапазоне от 75 000,00 денег до 7 700 000,00 денег. Как говорится "На любой вкус и кошелек"***

***Более глубокий анализ требует детального изучения. Однако, даже поверхностный взгляд на предоставленные данные, позволяет сделать вывод, что база данных содержит достаточно информации, позволяеющей расширить диапазон предложений для потенциальных покупателей  и подбора объекта. удовлетворяющего любой спрос.***

## ***Задание №4***
### ***Ответьте на несколько вопросов***

### ***4.1 Сколько в среднем стоят дома, у которых 2 спальни?***

In [16]:
f"Средняя стоимость дома с двумя спальнями составялет -{df[df['bedrooms'] == 2]['price'].mean().round(2)} денег"

'Средняя стоимость дома с двумя спальнями составялет -401372.68 денег'

### ***4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?***

In [17]:
f"В среднем общая площадь домов, у которых стоимость больше 600 000 денег-{df[df['price'] > 600000]['sqft_lot'].mean().round(2)} кв.футов."

'В среднем общая площадь домов, у которых стоимость больше 600 000 денег-20442.52 кв.футов.'

### ***4.3 Как много домов коснулся ремонт?***

In [18]:
f"Количество домов, в которых проведен ремонт - {df[df['yr_renovated'].astype('bool')].shape[0]}"

'Количество домов, в которых проведен ремонт - 914'

### ***4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?***

In [19]:
f"B среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4 на {df[df['grade'] > 10]['price'].mean().round(2) - df[df['grade'] < 4]['price'].mean().round(2)} денег, то есть в {df[df['grade'] > 10]['price'].mean().round(2) / df[df['grade'] < 4]['price'].mean().round(2)} раз."

'B среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4 на 1488885.12 денег, то есть в 8.846561897233203 раз.'

## ***Задание №4***
### ***Ответьте на несколько вопросов***

### ***Выберите дом клиенту***

### ***5.1 Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?***

In [20]:
df[(df['waterfront'] != 0) & (df['bathrooms'] >= 3) & (df['sqft_basement'] != 0)].shape[0]

41

### ***5.2 Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?***

In [21]:
choiсe = df[((df['view'] == 4) | (df['waterfront'] != 0)) & (df['yr_built'] >= 1980) & (df['condition'] == 5)]['price']
choiсe.min(), choiсe.max()

(1295000.0, 3000000.0)

### ***5.3 Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?***

In [22]:
df[(df['sqft_basement'] == 0) & (df['floors'] == 2) & (df['price'] < 150000)]['condition'].mean().round(2)

2.83